pygem/cad/ffd.py

```diff
@@ -0,0 +1,353 @@
+            # (actually, the WIRES) that define the bounds of the surface and
+            # TRIM the surface with them, to obtain the new face
+
+            # we start creating a face with the modified surface. we will
+            #later cut this new face with all the wires that the original
+            # face had this tolerance can be moved among the function
+            # parameters
+            brep = BRepBuilderAPI_MakeFace(occ_face, self.tolerance).Face()
+
+
+            # we here start looping on the wires of the original face
+            # in this first loop we do nothing but count the wires in this
+            # face few faces have more than one wire: if they have, it is
+            # because they have holes, and if this is the case one wire
+            # is the outer, and the others are the inner ones.
+            # the loop will also tell us which wire is the outer one
+            wire_count = 0
+            wire_explorer = TopExp_Explorer(face_aux, TopAbs_WIRE)
+            while wire_explorer.More():
+                wire = topods_Wire(wire_explorer.Current())
+                if wire == breptools_OuterWire(face_aux):
+                    print("Wire", wire_count+1, "is outer wire")
+                wire_count += 1
+                wire_explorer.Next()
+
+            #we now start really looping on the wires
+            #we will create a single curve joining all the edges in the wire
+            # the curve must be a bspline curve so we need to make conversions
+            # through all the way
+            wire_count = 0
+            outer_wires = []
+            inner_wires = []
+            brep_face = BRep_Tool.Surface(brep)
+            wire_explorer = TopExp_Explorer(face_aux, TopAbs_WIRE)
+            while wire_explorer.More():
+                wire = topods_Wire(wire_explorer.Current())
+                h_bspline_edge = GeomConvert_CompCurveToBSplineCurve()
+                edge_explorer = TopExp_Explorer(wire, TopAbs_EDGE)
+                edgesCount = 0
+                while edge_explorer.More():
+                    # performing some conversions to get the right format
+                    # (BSplineSurface)
+                    edge = topods_Edge(edge_explorer.Current())
+                    if not BRep_Tool.Degenerated(edge):
+                        bspline_converter = BRepBuilderAPI_NurbsConvert(edge)
+                        bspline_converter.Perform(edge)
+                        bspline_tshape_edge = bspline_converter.Shape()
+                        h_geom_edge, a, b = \
+                            BRep_Tool_Curve(topods_Edge(bspline_tshape_edge))
+                        this_bspline_edge = \
+                            geomconvert_CurveToBSplineCurve(h_geom_edge)
+                        bspline_geom_edge = this_bspline_edge
+                        h_bspline_edge.Add(this_bspline_edge, self.tolerance)
+                    edgesCount += 1
```

Comment:
Class private method (one underscore in front of the name e.g. `_foo()`) for performing the cast. Describe input and output variables.